src/NUnitFramework/framework/Internal/PreFilter.cs

```diff
@@ -0,0 +1,208 @@
+// Copyright (c) 2014-2018 Charlie Poole, Rob Prouse
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace NUnit.Framework.Internal
+{
+    /// <summary>
+    /// The PreFilter class implements a simplified filter for use in deciding which
+    /// Types and Methods should be used to generate tests. It is consructed with a 
+    /// list of strings, each of which may end up being interpreted in various ways.
+    /// </summary>
+    public class PreFilter
+    {
+        private static readonly char[] ARG_START = new char[] { '(', '<' };
+
+        private List<FilterElement> _filters = new List<FilterElement>();
+
+        /// <summary>
+        /// Return a new PreFilter, without elements, which is considered
+        /// empty and always matches.
+        /// </summary>
+        public static PreFilter Empty
+        {
+            get { return new PreFilter(); }
+        }
+
+        /// <summary>
+        /// Return true if the filer is empty, in which case it
+        /// always succeeds. Technically, this is just a filter and
+        /// you can add elements but it's best to use Empty when
+        /// you need an empty filter and new when you plan to add.
+        /// </summary>
+        public bool IsEmpty
+        {
+            get { return _filters.Count == 0; }
+        }
+
+        /// <summary>
+        /// Add a new filter element to the filter
+        /// </summary>
+        /// <param name="filterText"></param>
+        public void Add(string filterText)
+        {
+            _filters.Add(new FilterElement(filterText));
+        }
+
+        /// <summary>
+        /// Use the filter on a Type, returning true if the type matches the filter
+        /// and should therefore be included in the discovery process.
```

Comment:
Capitalizing `Type` looks odd to me. <see ref="Type"/> would render as `System.Type` but it would make just as much sense to me to use the common noun `type`. Could also start `Returns true if …`.  Same with the other overload.